contracts/upkeeps/Cron.sol

```diff
@@ -0,0 +1,278 @@
+
+  uint256 private s_nextCronJobID = 1;
+  uint256[] private s_activeCronJobIDs;
+
+  mapping(uint256 => uint256) private s_nextTicks;
+  mapping(uint256 => Spec) private s_specs;
+  mapping(uint256 => address) private s_targets;
+  mapping(uint256 => bytes) private s_handlers;
+  mapping(uint256 => bytes32) private s_handlerSignatures;
+
+  /**
+   * @notice Get the id of an eligible cron job
+   * @return upkeepNeeded signals if upkeep is needed, performData is an abi encoding
+   * of the id and "next tick" of the elligible cron job
+   */
+  function checkUpkeep(bytes calldata)
+    external
+    view
+    override
+    whenNotPaused
+    returns (bool, bytes memory)
+  {
+    // DEV: start at a random spot in the list so that checks are
+    // spread evenly among cron jobs
+    uint256 idx;
+    uint256 id;
+    uint256 numCrons = s_activeCronJobIDs.length;
+    uint256 startIdx = block.number % numCrons;
+    for (idx = startIdx; idx < numCrons; idx++) {
+      id = s_activeCronJobIDs[idx];
+      if (block.timestamp >= s_nextTicks[id]) {
+        return (
+          true,
+          abi.encode(id, s_specs[id].nextTick(), s_targets[id], s_handlers[id])
+        );
+      }
+    }
+    for (idx = 0; idx < startIdx; idx++) {
+      id = s_activeCronJobIDs[idx];
+      if (block.timestamp >= s_nextTicks[id]) {
+        return (
+          true,
+          abi.encode(id, s_specs[id].nextTick(), s_targets[id], s_handlers[id])
+        );
+      }
+    }
+    return (false, bytes(""));
+  }
+
+  /**
+   * @notice Executes the cron job with id encoded in performData
+   * @param performData abi encoding of cron job ID and the cron job's next run-at datetime
+   */
+  function performUpkeep(bytes calldata performData)
+    external
+    override
+    onlyWhitelist
+    whenNotPaused
+  {
+    (uint256 id, uint256 nextTick, address target, bytes memory handler) = abi
+      .decode(performData, (uint256, uint256, address, bytes));
+    if (block.timestamp < s_nextTicks[id]) {
+      revert PerformanceNotNeeded();
+    }
+    if (handlerSig(target, handler) != s_handlerSignatures[id]) {
+      revert InvalidHandler();
+    }
+    s_nextTicks[id] = nextTick;
+    (bool success, bytes memory payload) = target.call(handler);
+    if (!success) {
+      revert CallFailed(id, getRevertMsg(payload));
+    }
+    emit CronJobExecuted(id, block.timestamp);
+  }
+
+  // SETTERS
+
+  /**
+   * @notice Creates a cron job from the given encoded spec. The nextTick is calculated
+   * as part of the tx (expensive!)
+   * @param target the destination contract of a cron job
+   * @param handler the function signature on the target contract to call
+   * @param encodedCronSpec abi encoding of a cron spec
+   */
+  function createCronJob(
+    address target,
+    bytes memory handler,
+    bytes memory encodedCronSpec
+  ) external {
+    Spec memory spec = abi.decode(encodedCronSpec, (Spec));
+    createCronJobFromSpec(target, handler, spec, spec.nextTick());
```

Comment:
What do you think of exposing `createCronJobFromSpec` so that users can avoid that gas cost and pass in a start time if they'd like. (Could pass 0 and just let it just immediately tick.)